Sources/AblyChat/Subscription.swift

```diff
@@ -4,22 +4,112 @@
 //
 // At some point we should define how this thing behaves when you iterate over it from multiple loops, or when you pass it around. I’m not yet sufficiently experienced with `AsyncSequence` to know what’s idiomatic. I tried the same thing out with `AsyncStream` (two tasks iterating over a single stream) and it appears that each element is delivered to precisely one consumer. But we can leave that for later. On a similar note consider whether it makes a difference whether this is a struct or a class.
 //
-// TODO: I wanted to implement this as a protocol (from which `MessageSubscription` would then inherit) but struggled to do so, hence the struct. Try again sometime. We can also revisit our implementation of `AsyncSequence` if we migrate to Swift 6, which adds primary types and typed errors to `AsyncSequence` and should make things easier.
-public struct Subscription<Element>: Sendable, AsyncSequence {
-    // This is a workaround for the fact that, as mentioned above, `Subscription` is a struct when I would have liked it to be a protocol. It allows people mocking our SDK to create a `Subscription` so that they can return it from their mocks. The intention of this initializer is that if you use it, then the created `Subscription` will just replay the sequence that you pass it.
-    public init<T: AsyncSequence>(mockAsyncSequence _: T) where T.Element == Element {
-        fatalError("Not implemented")
+// TODO: I wanted to implement this as a protocol (from which `MessageSubscription` would then inherit) but struggled to do so, hence the struct. Try again sometime. We can also revisit our implementation of `AsyncSequence` if we migrate to Swift 6, which adds primary types and typed errors to `AsyncSequence` and should make things easier; see https://github.com/ably-labs/ably-chat-swift/issues/21.
+public struct Subscription<Element: Sendable>: Sendable, AsyncSequence {
+    private enum Mode: Sendable {
+        case `default`(stream: AsyncStream<Element>, continuation: AsyncStream<Element>.Continuation)
+        case mockAsyncSequence(TypeErasedNonThrowingAsyncSequence)
```

Comment:
What does "TypeErased" mean?

Reply:
A _type-erased_ type is a type `S` that:

1. implements some protocol `P`
2. by wrapping another value of some arbitrary type `T` that implements `P`
3. without exposing `T` in the type of `S`

Concretely, here, `TypeErasedAsyncSequence` is a type that implements a `AsyncSequence` with element type `Element`, by wrapping the type `T` that’s passed to `TypeErasedAsyncSequence.init(asyncSequence:)`, without exposing the type `T`. Why do we want to do this? It’s so that we don’t have to add a second generic type parameter to `Subscription` to describe the type that’s passed to `Subscription.init(mockAsyncSequence)`.

Functionally, this type is similar to the `any` keyword that you may have seen used in Swift; e.g. `any Collection<String>`. Ideally, we’d use the same syntax here, i.e. write `init(mockAsyncSequence: any AsyncSequence<T, Never>)`, but this isn’t possible in Swift 5 because `AsyncSequence` doesn’t have primary associated types. In Swift 6, you can write this.

Reply:
Ok, but I still don't understand why engrave this implementation detail into the type name. It's confusing for reader. If it's similar to `any` why can't it be `AnyNonThrowingAsyncSequence` or just `NonThrowingAsyncSequence`?

Reply:
I think that `AnyNonThrowingAsyncSequence` is a better name; will change it!

Reply:
(that's the naming convention used for type-erased types by Apple; see e.g. [`AnyView`](https://developer.apple.com/documentation/swiftui/anyview).)

Reply:
Oh, so it's an Apple concept.. okay than. Still, yes, I think `AnyNonThrowingAsyncSequence` serves this well.